src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -68,6 +68,36 @@ AbstractLQPNode::~AbstractLQPNode() {
   if (_inputs[1]) _inputs[1]->_remove_output_pointer(*this);
 }
 
+size_t AbstractLQPNode::hash() const {
+  auto hash = boost::hash_value(static_cast<LQPNodeType>(type));
+  for (const auto& expression : node_expressions) {
+    boost::hash_combine(hash, expression->hash());
+  }
+  boost::hash_combine(hash, _on_hash());
+
+  switch (type) {
```

Comment:
Abstract base classes shouldn't know about derived types.

```c++
if (left_input) hash_combine(hash, left_input->hash());
if (right_input) hash_combine(hash, right_input->hash());
```

Here too, visit_lqp would avoid redundant traversal of subplans below diamond shapes